perf(search): improve fuzzy character search performance by ~13x (4.5s → 350ms)

joenunezb opened the pull request:
# Problem

The character search feature is _extremely_ slow when using **fuzzy** matching. It was taking upwards of 5 seconds to search anything which hinders useability.

# Solution

The issue is that the search algorithm calls out to make a fuzzy search multiple times in an effort to filter various criteria out. On each of these calls, Fuse is re-initialized up to 40+ times. Instead of re-initializing fuse, simply cache the fuse results.

In order to avoid stale cache hits, we only use the cache within the core search code, any other code path just clears the cache. 

This is mainly due to any other approaches resulting in either a leaky abstraction, where code paths unrelated to search had to manually clear the cache which is error prone OR a cost to performance if we wanted to hash all the inputs to the cache given that the data sources do not have a `lastUpdated` timestamp that we can check quickly to invalidate the cache.


# Test

## Before fix:

Search used to vary between 3-5 seconds depending on tags and character info

![before_fuzzy_search_fix](https://github.com/user-attachments/assets/9a03985d-71e3-4ed4-9c47-80f93a9cade6)


## After fix:

Search varies between 150ms-350ms

![after_fuzzy_search_fix](https://github.com/user-attachments/assets/ac7108ea-c0e0-47c9-a87e-20028eb409b7)


I also searched through World Info, Personas, Group Adding, and did not find any regressions. 


## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

joenunezb committed:
Sped up character search by 93%

joenunezb committed:
Handle stale cache

joenunezb committed:
Remove duplicated code

joenunezb committed:
Coment out log statements

Cohee1207 commented on public/scripts/power-user.js:
```diff
@@ -1825,64 +1841,111 @@ async function loadContextSettings() {
+/**
+ * Clears all fuzzy search caches
+ */
+export function clearFuzzySearchCaches() {
```
It would be a good idea to call this function in places where objects stored in resultMaps are discarded (i.e. methods that load characters, groups, etc. from the backend). Otherwise this would leak memory as object references are still exist in the Map, while being discarded from the main arrays.

Otherwise, take a look at Weak collections, such as WeakSet to store cached results instead.

joenunezb replied:
Thank you Cohee! I have no idea what I am doing when it comes to js haha. Let me take a look at Weak collections.

Regarding memory leaks, as long as the application is running search is accessible correct? Say we load a character or a group, shouldn't the cache get invalidated if something changes and otherwise shouldn't the cache remain available to increase search performance on the next search?

I'm just trying to understand the use case in which storing this cache as long as the application is running wouldn't be beneficial.

Cohee1207 replied:
The search results cache holds references to objects. A reload of characters/groups/other entities discards the previously stored values which causes the garbage collection to remove them from the memory and populate with newly acquired values, as long as the reference is not help by any other object. By storing them in a regular map, you make search results use stale objects, and prevent them from being garbage collected.

joenunezb replied:
Looked into Weak collections and they seem like they might not properly work. Would you mind pointing me towards where I can find the loads from backend so I can properly reset the cache? Otherwise I can introduce an LRU cache, maybe that might limit memory and hopefully keep the perf improvement.

Cohee1207 replied:
Another approach would be timed cache, i.e. to automatically discard search results if they are ~1 minute old. But you'd have to do some due diligence with the implementation details.

Cohee1207 replied:
Here's the definition for getCharacters.

https://github.com/SillyTavern/SillyTavern/blob/release/public/script.js#L1692

joenunezb replied:
I placed the call to clear cache on server load but the cache wasn't being cleared often unless we performed an operation that would make the request. 

I've instead placed it at the end of `PrintCharacters` which is called in many places due to it being wrapped by `PrintCharactersDebounced`.

![after_fix_with_cache_clearing](https://github.com/user-attachments/assets/e69c97f1-523f-4470-adfe-d55b3f29f79c)
![after_fix_with_cache_clearing_logs](https://github.com/user-attachments/assets/73e4f7b7-9b90-47ba-99e6-a5e8520fc398)

The performance improvement is still there, increasing now 200ms-360ms but it's still manageable.

I've also removed the key hashing since we're explicitly clearing the cache after printing the characters which will reinitialize it on every search @Wolfsblvt so that should be a more general catch-all to the concern you raised.

Let me know what you guys think.

Cohee1207 replied:
Are you sure about placing it in `printCharacters`? It's called every time in order to update the list in the right sidebar. So every time the search is completed and the list is rendered, the cache would be dropped. I was thinking about `getCharacters`, in the middle between `getGroups()` and `printCharacters()` calls.

joenunezb replied:
Adding it to `printCharacters` allows it to be called in various places, for example, whenever a tag is added or dropped removing the need to hash the keys for the cache. This reduces a potential key hash collision which wasn't handled before (even though extremely unlikely). 

Rebuilding the cache the first time is taking < 350ms per search, the core issue before was that this was being done every single time which led to "death by thousands cuts" and accumulating to ~4-5s per search. If we rebuild the cache during the first search and re-use it and then clear it by the end of the search it essentially leads to the same performance improvement after removing the key hashing.

If we do not add it here, I'll have to add the key hashing again (which I don't mind if you believe that's best) and I can add back the key invalidation after `getGroups` and `printCharacters` calls.

joenunezb replied:
As a side note, I went back and realized I had a bug with hashing the keys as the cache was not being properly updated. I needed to hash the values not the keys but that introduced additional latency.

The search time went from ~350ms to ~800ms with hashing the values. Maybe a bug on my end but it seems that invalidating the cache on `printCharacters` is likely the most performant, and less error prone (given that it gets reset at search time). 

For fun I checked what performance we get if we had some other mechanism to know when characters are updated (i.e calling this on character data save if tags are updated or description is updated) and it's ~150-200ms which looks really fast haha. The only issue is that if we distribute the cache clearing to other areas, it sort of turns into a leaky abstraction given that this is mainly a search cache and now other parts of the system need to know how to reset it. 

The most ideal would be having a `lastUpdated` area in any of our data sources (tags, groups, worldInfo, characters) which then we can invalidate the cache by simply checking if cached lastUpdated < current lastUpdated.

joenunezb replied:
Disregard, added a comment on new flow.

Wolfsblvt commented:
Does this also take in mind that for quite a few of the elements, the data relevant for the search can change, without the element being added/deleted?

For example, characters having new tags assigned or tags deleted, which will change the search results.

I fear introducing caching for something as important as search results should be done really carefully. Cache invalidation is a really hard topic. High chances to get it wrong.

joenunezb commented:
> Does this also take in mind that for quite a few of the elements, the data relevant for the search can change, without the element being added/deleted?
> 
> For example, characters having new tags assigned or tags deleted, which will change the search results.
> 
> I fear introducing caching for something as important as search results should be done really carefully. Cache invalidation is a really hard topic. High chances to get it wrong.

Hey! I totally understand, the two hardest things, naming and caching right haha? Yes, I tested this by adding and removing a tag to a character and it successfully invalidated the cache and updated it. The main purpose of introducing the hashing of the lookup keys are exactly for this use case in which the search gets modified by the specific lookup values similarity search works on. If any of them change the hash will change.

I commented out the logs (logging itself takes ~1-3ms) but I can post log results if you're interested or make a recording.

joenunezb committed:
Remove key hashing and explicitly clear the cache after printing characters

joenunezb committed:
Remove hash function

Cohee1207 commented on public/script.js:
```diff
@@ -1515,6 +1516,7 @@ export async function printCharacters(fullRefresh = false) {
     });
 
     favsToHotswap();
+    clearFuzzySearchCaches();
```
I suppose similar calls should also be made in places where personas and world info are reloaded to avoid stale caches.

https://github.com/SillyTavern/SillyTavern/blob/release/public/scripts/personas.js#L158
https://github.com/SillyTavern/SillyTavern/blob/release/public/scripts/world-info.js#L1571

joenunezb replied:
Thank you! Will add

joenunezb replied:
I decided against manually clearing the caches for a more encapsulated approach via filters.js

joenunezb committed:
Document Fuzzy Search process

joenunezb commented on docs/fuzzy_search_flow.md:
```diff
@@ -0,0 +1,113 @@
+    GEL-->>PC: Return final entities list
+    
+    PC->>Cache: clearFuzzySearchCaches()
+    Note over Cache: Cache is cleared at the end of<br/>each printCharacters call,<br/>ensuring fresh results for next search
```
@Cohee1207 I think having this in `printCharacters` addresses most if not all cache related issues due to this flow. I did not know how else to really explain my thought process so created `docs/`. If you feel like there's a more appropriate place or would like me to remove let me know.

joenunezb replied:
Disregard

joenunezb committed:
Attempt to allow preview for github

joenunezb committed:
Merge branch 'staging' of https://github.com/joenunezb/SillyTavern into optimize/improve-search

joenunezb committed:
Encapsulate logic into filters instead of spreading around

joenunezb committed:
Remove unnecesary key only clearing for cache clearing

joenunezb committed:
Account for optional cache and remove timing code

joenunezb commented:
@Cohee1207 I decided to rethink how to avoid manually clearing the cache everywhere. It's really error prone as @Wolfsblvt was mentioning given that it may be fine for now but someone may modify/change some code leading to the cache invalidation logic from not triggering.

Since the main user of the fuzzy searching is filters.js I moved the core logic there. Now the fuzzy search functions will utilize + update the cache if provided one otherwise will simply work as before.

In addition, we default clear the cache every time `applyFilters` is called, and only avoid clearing it for specific nested search logic that happens within `getEntitiesList`.

This means that all other code will behave as if the caching mechanism is not there, reducing the chance of regression and manual cache invalidation along with optimizing the nested function calling behavior that originally led to the 5s or so in search time. This should also optimize any other area where `getEntitiesList` was being called before too.

Performance remains the same per seach (~350ms):
![after_fix_filters](https://github.com/user-attachments/assets/a3adc258-0618-4de4-a67e-a9a666e80594)


I've also tested removing/adding tags, WiSearch, and Persona search all working as expected. 


As I mentioned in the updated description and a previous comment, in order to get that sweet < 150ms search, I would need a field added to all input data sources `lastUpdateTime` that will allow me to check the cached `lastUpdateTime` vs current `lastUpdateTime` so that I am able to not invalidate the cache unless something has changed.

I tracked the character metadata/information to a schema which I believe may cause issues if I modified willy nilly so I avoided making any changes for now. Maybe in a follow up PR I can take a step at it with some guidance.

Cohee1207 reviewed (approved):
Tested. I wasn't able to trigger any stale cache hits, and performance improvements seems nice. Thanks.
